### PR TITLE
feat: support precompute partition for iceberg

### PR DESCRIPTION
**Pull request by ZENOTME**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This is the final part of #13898:
1. support get partition info at frontend 
2. add the project node to compute the partition value and exchange node to shuffle by it 

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Review by liurenjie1024 — COMMENTED** [review]
Generally LGTM, thanks!

**Comment by BugenZhao** [comment]
Is this PR specific to Iceberg? Would you please update the PR title to make it more detailed?

**Comment by ZENOTME** [comment]
> Is this PR specific to Iceberg? Would you please update the PR title to make it more detailed?

Sorry. Update it.

File: proto/stream_plan.proto
Comment (ZENOTME) [reply proto/stream_plan.proto:226]: Maybe we can use a special column name `_rw_partition` and search for it in the whole schema or the last column.🤔

File: src/connector/src/sink/iceberg/mod.rs
Comment (ZENOTME) [reply src/connector/src/sink/iceberg/mod.rs:381]: Because we need to reuse this function in frontend to extract the partition spec of a table. 
https://github.com/risingwavelabs/risingwave/blob/0cc5ba92ce82b352d7195d1056c4a1dadb5edb8f/src/frontend/src/handler/create_sink.rs#L324
We can make the `create_table` pub only.

File: proto/stream_plan.proto
Comment (ZENOTME) [reply proto/stream_plan.proto:226]: Seems the project can't support specifying a field name now.

Comment (liurenjie1024) [reply proto/stream_plan.proto:226]: I see, let's keep it for now. Please add some comments for this field.

File: src/frontend/src/optimizer/plan_node/stream_sink.rs
Comment (ZENOTME) [reply src/frontend/src/optimizer/plan_node/stream_sink.rs:239]: I find it inconvenient to test it for Iceberg because I need to connect the catalog and get related info. 
We test the sink using `handle_explain`, which means that there isn't a place to insert the mock partition info in plan test.
- One way to test it is to mock the partition info in `blackhole` connector. But this will add the code for test in frontend, I'm not sure whether it's a good way.
```
pub async fn get_partition_compute_info(
    with_options: &WithOptions,
) -> Result<Option<PartitionComputeInfo>> {
   match connector.as_str() {
        ICEBERG_SINK => {
            ..
        }
       BLACK_HOLE => { ..//generate mock info }
        _ => Ok(None),
    }
}
```
- Another way is to modify the test way for sink, maybe we can use create_sink instead of test using explain and add a param to pass the partition info so that we can mock the partition info outside. 
https://github.com/risingwavelabs/risingwave/blob/ee08ac5cbd9e2f1042ea8b61a7c9958d81840823/src/frontend/planner_test/src/lib.rs#L550

Comment (liurenjie1024) [reply src/frontend/src/optimizer/plan_node/stream_sink.rs:239]: > I find it inconvenient to test it for Iceberg because I need to connect the catalog and get related info.

How about creating a mocked iceberg catalog?

File: proto/stream_plan.proto
Comment (ZENOTME) [reply proto/stream_plan.proto:226]: Add a comment at https://github.com/risingwavelabs/risingwave/blob/33187d8a8f51962931d15cfe3839b3a8ac00dcad/src/connector/src/sink/mod.rs#L276. 
and (use a ref for consistency 
https://github.com/risingwavelabs/risingwave/blob/b46fa781e31eed9c38a920bbd3ab1ed09a73b520/src/connector/src/sink/catalog/desc.rs#L72

File: src/frontend/planner_test/tests/testdata/output/sink.yaml
Comment (liurenjie1024) [inline src/frontend/planner_test/tests/testdata/output/sink.yaml:175]: This test cases seem not enough, in the rfc there are 4 cases, please add test for each case.

File: src/frontend/src/handler/create_sink.rs
Comment (tabVersion) [inline src/frontend/src/handler/create_sink.rs:336]: so we are changing the distribution key to adjust each CN's hash range?

**Review by tabVersion — COMMENTED** [review]
not familiar with the context, waiting for renjie's approval

File: src/frontend/src/optimizer/plan_node/stream_sink.rs
Comment (ZENOTME) [reply src/frontend/src/optimizer/plan_node/stream_sink.rs:58]: Yes, seems delta also has partition concept. I guess it can apply it.🤔

File: src/frontend/src/handler/create_sink.rs
Comment (ZENOTME) [reply src/frontend/src/handler/create_sink.rs:336]: Yes, we can see that for sparse partition, we will add the exchange before sink.
https://github.com/risingwavelabs/risingwave/blob/8b263cb9a4471386f05ddd0239440a4a54ece73b/src/frontend/planner_test/tests/testdata/output/sink.yaml#L233

**Review by liurenjie1024 — APPROVED** [review]
LGTM, thanks!

**Comment by liurenjie1024** [comment]
cc @xxchan @BugenZhao PTAL for cargo.lock

**Review by hzxa21 — APPROVED** [review]
LGTM for Cargo.lock

**Pull request merged**
